### PR TITLE
fix: layout grid bug fixes, improvements, refactoring

### LINKED ISSUE #2612: Negative margin values can result in form layout grid overflowing parent container
Negative margin values can result in form layout grid overflowing parent container.  This can cause problems for example when using the form layout in a dialog - the form layout is actually bigger than the dialog causing the dialog to have an unnecessary scroll bar.  We need to find an alternative to the negative values

Screen Shot 2021-07-27 at 4 18 36 PM

### PR DESCRIPTION
**Pull request by platon-rov**
## Related Issue

Closes #2612.

## Description

Negative margins were applied to the inappropriate elements that make rows wider than container and cause appearing the scrollbars. 

Gutter between rows was set by padding for the column not by a margin for the row.

### Before + After:

Screenshot 2021-09-09 at 21 03 43

#### Please check whether the PR fulfills the following requirements

1. The output matches the design specs
- [X] All values are in `rem`
- [X] Text elements follow the truncation rules
- [X] hover state of the element follow design spec
- [X] focus state of the element follow design spec
- [X] active state of the element follow design spec
- [X] selected state of the element follow design spec
- [X] selected hover state of the element follow design spec
- [X] pressed state of the element follow design spec
- [X] Responsiveness rules - the component has modifier classes for all breakpoints
- [X] Includes Compact/Cosy/Tablet design
- [X] RTL support
2. The code follows fundamental-styles code standards and style
- [X] only one top level `fd-*` class is used in the file
- [X] BEM naming convention is used
- [X] Mixins are used for repeatable code (`fd-rtl`, `fd-ellipsis`, `fd-flex`, `fd-selected`, `fd-focus`, ect.)
- [X] A11y support - keyboard support, screenreader support, proper ARIA attributes, etc.
- [X] `fd-reset()` mixin is applied to all elements
- [X] Variables are used, if some value is used more than twice.
- [X] Checked if current components can be reused, instead of having new code.
3. Testing
- [X] tested Storybook examples with "CSS Resources" `normalize` option 
- [X] tested Storybook examples with "CSS Resources" `unnormalize` option 
- [X] Verified all styles in IE11
- [x] Updated tests
- [X] last commit message should have `[ci visual]` so it can trigger chromatic visual regression (e.g. `test: run chromatic visual regression [ci visual]`)
4. Documentation
- [X] Storybook documentation has been created/updated
- [X] Breaking Changes [wiki](https://github.com/SAP/fundamental-styles/wiki/Breaking-Changes) has been updated in case of breaking changes.

### CODE REVIEW

**Comment by InnaAtanasova** [comment]
@platon-rov please add `before` and `after` screenshots of a component affected by the changes.

File: src/layout-grid.scss [hunk 1]
```diff
@@ -52,31 +49,37 @@ $cols: 12;
...
-  margin: 0 -0.25rem 0 -0.25rem;
+  padding: 0 0.25rem;
 
-  &--no-gap {
-    margin: 0;
+  &.#{$blockContainer}--no-horizontal-gap,
+  &.#{$blockContainer}--no-gap {
+    .#{$blockRow} {
+      margin-left: 0;
+      margin-right: 0;
```
Comment (InnaAtanasova) [inline src/layout-grid.scss:69]: There's a mixin for resetting the horizontal margin:
```
@mixin fd-reset-margins-x() {
  margin-left: 0;
  margin-right: 0;
}
```

File: src/layout-grid.scss [hunk 2]
```diff
@@ -52,31 +49,37 @@ $cols: 12;
...
 
-    .#{$block} {
-      padding: 0;
+      .#{$blockCol} {
+        padding-left: 0;
+        padding-right: 0;
```
Comment (InnaAtanasova) [inline src/layout-grid.scss:73]: Please use the mixin:
```
@mixin fd-reset-paddings-x() {
  padding-left: 0;
  padding-right: 0;
}
```

File: src/layout-grid.scss [hunk 3]
```diff
@@ -85,129 +88,76 @@ $cols: 12;
...
+  &:first-child {
+    margin-top: -0.5rem;
   }
 
-  .#{$block} {
-    .#{$blockRow} + .#{$blockRow} {
-      margin-top: -0.5rem;
-    }
+  .#{$blockCol} {
+    padding-top: 0.5rem;
```
Comment (InnaAtanasova) [inline src/layout-grid.scss:103]: could you please put values like `0.5rem`, `0.25rem`, etc. in variables. This way we can easily change and maintain them.

**Review by InnaAtanasova — CHANGES_REQUESTED** [review]

**Review by InnaAtanasova — APPROVED** [review]

**Review by Smiranin — APPROVED** [review]

**Pull request merged**